File: sources/Metal/Renderers/iTermMarginRenderer.m
//
//  iTermMarginRenderer.m
//  iTerm2SharedARC
//
//  Created by George Nachman on 11/19/17.
//

#import "iTermMarginRenderer.h"

#import "FutureMethods.h"
#import "iTermMetalBufferPool.h"
#import "iTermShaderTypes.h"

NS_ASSUME_NONNULL_BEGIN

@implementation iTermMarginRendererTransientState
@end

@implementation iTermMarginRenderer {
    iTermMetalCellRenderer *_blendingRenderer;
#if ENABLE_TRANSPARENT_METAL_WINDOWS
    iTermMetalCellRenderer *_nonblendingRenderer NS_AVAILABLE_MAC(10_14);
    iTermMetalCellRenderer *_compositeOverRenderer NS_AVAILABLE_MAC(10_14);
#endif
    iTermMetalBufferPool *_colorPool;
    iTermMetalBufferPool *_verticesPool;
    iTermMetalBufferPool *_deselectedVerticesPool;
}

- (nullable instancetype)initWithDevice:(id<MTLDevice>)device {
    self = [super init];
    if (self) {
#if ENABLE_TRANSPARENT_METAL_WINDOWS
        if (iTermTextIsMonochrome()) {
            _nonblendingRenderer = [[iTermMetalCellRenderer alloc] initWithDevice:device
                                                               vertexFunctionName:@"iTermMarginVertexShader"
                                                             fragmentFunctionName:@"iTermMarginFragmentShader"
                                                                         blending:nil
                                                                   piuElementSize:0
                                                              transientStateClass:[iTermMarginRendererTransientState class]];
            _compositeOverRenderer = [[iTermMetalCellRenderer alloc] initWithDevice:device
                                                                 vertexFunctionName:@"iTermMarginVertexShader"
                                                               fragmentFunctionName:@"iTermMarginFragmentShader"
                                                                           blending:[iTermMetalBlending premultipliedCompositing]
                                                                     piuElementSize:0
                                                                transientStateClass:[iTermMarginRendererTransientState class]];
        }
#endif
        _blendingRenderer = [[iTermMetalCellRenderer alloc] initWithDevice:device
                                                        vertexFunctionName:@"iTermMarginVertexShader"
                                                      fragmentFunctionName:@"iTermMarginFragmentShader"
                                                                  blending:[[iTermMetalBlending alloc] init]
                                                            piuElementSize:0
                                                       transientStateClass:[iTermMarginRendererTransientState class]];
        _colorPool = [[iTermMetalBufferPool alloc] initWithDevice:device bufferSize:sizeof(vector_float4)];
        _verticesPool = [[iTermMetalBufferPool alloc] initWithDevice:device bufferSize:sizeof(vector_float2) * 6 * 6];
        _deselectedVerticesPool = [[iTermMetalBufferPool alloc] initWithDevice:device bufferSize:sizeof(vector_float2) * 6 * 8];
    }
    return self;
}

- (iTermMetalFrameDataStat)createTransientStateStat {
    return iTermMetalFrameDataStatPqCreateMarginTS;
}

- (iTermMetalCellRenderer *)rendererForConfiguration:(iTermCellRenderConfiguration *)configuration {
#if ENABLE_TRANSPARENT_METAL_WINDOWS
    if (iTermTextIsMonochrome()) {
        if (configuration.hasBackgroundImage) {
            return _compositeOverRenderer;
        } else {
            return _nonblendingRenderer;
        }
    }
#endif
    return _blendingRenderer;
}

- (void)drawWithFrameData:(nonnull iTermMetalFrameData *)frameData
           transientState:(__kindof iTermMetalRendererTransientState *)transientState {
    iTermMarginRendererTransientState *tState = transientState;
    if (tState.hasSelectedRegion) {
        [self drawWithSelectedRegionHavingFrameData:frameData tState:tState];
    } else {
        [self drawRegularHavingFrameData:frameData tState:tState];
    }
}

- (void)drawRegularHavingFrameData:(nonnull iTermMetalFrameData *)frameData
                            tState:(nonnull iTermMarginRendererTransientState *)tState {
    [self initializeRegularVertexBuffer:tState];
    vector_float4 regularColor = tState.regularColor;
    if (iTermTextIsMonochrome()) {
        regularColor.x *= regularColor.w;
        regularColor.y *= regularColor.w;
        regularColor.z *= regularColor.w;
    }
    id<MTLBuffer> colorBuffer = [_colorPool requestBufferFromContext:tState.poolContext
                                                           withBytes:&regularColor
                                                      checkIfChanged:YES];
    iTermMetalCellRenderer *cellRenderer = [self rendererForConfiguration:tState.cellConfiguration];
    [cellRenderer drawWithTransientState:tState
                           renderEncoder:frameData.renderEncoder
                        numberOfVertices:6 * 4
                            numberOfPIUs:0
                           vertexBuffers:@{ @(iTermVertexInputIndexVertices): tState.vertexBuffer }
                         fragmentBuffers:@{ @(iTermFragmentBufferIndexMarginColor): colorBuffer }
                                textures:@{}];
}

// Matches function in iTermBackgroundColor.metal
static vector_float4 iTermBlendColors(vector_float4 src, vector_float4 dst) {
    vector_float4 out;
    out.w = src.w + dst.w * (1 - src.w);
    if (out.w > 0) {
        out.xyz = (src.xyz * src.w + dst.xyz * dst.w * (1 - src.w)) / out.w;
    } else {
        out.xyz = 0;
    }
    return out;
}

- (void)drawWithSelectedRegionHavingFrameData:(nonnull iTermMetalFrameData *)frameData
                                       tState:(nonnull iTermMarginRendererTransientState *)tState {
    // Do up to 8 separate draws
    iTermMetalCellRenderer *cellRenderer = [self rendererForConfiguration:tState.cellConfiguration];

    int n = [self initializeDeselected:NO vertexBuffer:tState];
    [self drawWithRenderEncoder:frameData.renderEncoder
                          color:tState.regularColor
                    poolContext:tState.poolContext
                   cellRenderer:cellRenderer
                   vertexBuffer:tState.vertexBuffer
                       numQuads:n
                 transientState:tState];

    const NSUInteger suppressedPx = tState.suppressedBottomHeight * tState.cellConfiguration.scale;
    if (tState.suppressedBottomHeight > 0) {
        // Avoid drawing margin in suppressed region. Background color renderer will handle it for us.
        MTLScissorRect scissorRect = {
            .x = 0,
            .y = 0,
            .width = tState.cellConfiguration.viewportSize.x,
            .height = tState.cellConfiguration.viewportSize.y - suppressedPx
        };
        [frameData.renderEncoder setScissorRect:scissorRect];
    }

    n = [self initializeDeselected:YES vertexBuffer:tState];
    const vector_float4 blendedColor = iTermBlendColors(tState.deselectedColor, tState.regularColor);
    [self drawWithRenderEncoder:frameData.renderEncoder
                          color:blendedColor
                    poolContext:tState.poolContext
                   cellRenderer:cellRenderer
                   vertexBuffer:tState.vertexBuffer
                       numQuads:n
                 transientState:tState];
    if (tState.suppressedBottomHeight > 0) {
        MTLScissorRect scissorRect = {
            .x = 0,
            .y = 0,
            .width = tState.cellConfiguration.viewportSize.x,
            .height = tState.cellConfiguration.viewportSize.y
        };
        [frameData.renderEncoder setScissorRect:scissorRect];
    }
}

- (void)drawWithRenderEncoder:(id<MTLRenderCommandEncoder>)renderEncoder
                        color:(vector_float4)nonPremultipliedColor
                  poolContext:(iTermMetalBufferPoolContext *)poolContext
                 cellRenderer:(iTermMetalCellRenderer *)cellRenderer
                 vertexBuffer:(id<MTLBuffer>)vertexBuffer
                     numQuads:(int)numQuads
               transientState:(iTermMetalRendererTransientState *)transientState {
    if (numQuads == 0) {
        return;
    }
    vector_float4 color = nonPremultipliedColor;
    if (iTermTextIsMonochrome()) {
        color.x *= color.w;
        color.y *= color.w;
        color.z *= color.w;
    }
    id<MTLBuffer> colorBuffer = [_colorPool requestBufferFromContext:poolContext
                                                           withBytes:&color
                                                      checkIfChanged:YES];

    [cellRenderer drawWithTransientState:transientState
                           renderEncoder:renderEncoder
                        numberOfVertices:6 * numQuads
                            numberOfPIUs:0
                           vertexBuffers:@{ @(iTermVertexInputIndexVertices): vertexBuffer }
                         fragmentBuffers:@{ @(iTermFragmentBufferIndexMarginColor): colorBuffer }
                                textures:@{}];
}

- (BOOL)rendererDisabled {
    return NO;
}

- (nullable __kindof iTermMetalRendererTransientState *)createTransientStateForCellConfiguration:(nonnull iTermCellRenderConfiguration *)configuration
                                                                          commandBuffer:(nonnull id<MTLCommandBuffer>)commandBuffer {
    iTermMetalCellRenderer *renderer = [self rendererForConfiguration:configuration];
    __kindof iTermMetalRendererTransientState * _Nonnull transientState =
        [renderer createTransientStateForCellConfiguration:configuration
                                             commandBuffer:commandBuffer];
    return transientState;
}

- (vector_float2 *)appendVerticesForQuad:(CGRect)quad vertices:(vector_float2 *)v {
    *(v++) = simd_make_float2(CGRectGetMaxX(quad), CGRectGetMinY(quad));
    *(v++) = simd_make_float2(CGRectGetMinX(quad), CGRectGetMinY(quad));
    *(v++) = simd_make_float2(CGRectGetMinX(quad), CGRectGetMaxY(quad));

    *(v++) = simd_make_float2(CGRectGetMaxX(quad), CGRectGetMinY(quad));
    *(v++) = simd_make_float2(CGRectGetMinX(quad), CGRectGetMaxY(quad));
    *(v++) = simd_make_float2(CGRectGetMaxX(quad), CGRectGetMaxY(quad));

    return v;
}

- (void)initializeRegularVertexBuffer:(iTermMarginRendererTransientState *)tState {
    CGSize size = CGSizeMake(tState.configuration.viewportSize.x,
                             tState.configuration.viewportSize.y);
    const NSEdgeInsets margins = tState.margins;
    vector_float2 vertices[6 * 4];
    vector_float2 *v = &vertices[0];
    // Top
    v = [self appendVerticesForQuad:CGRectMake(0,
                                               0,
                                               size.width,
                                               margins.top)
                           vertices:v];

    // Bottom
    v = [self appendVerticesForQuad:CGRectMake(0,
                                               size.height - margins.bottom,
                                               size.width,
                                               margins.bottom)
                           vertices:v];

    const CGFloat innerHeight = size.height - margins.bottom - margins.top;

    // Left
    v = [self appendVerticesForQuad:CGRectMake(0,
                                               margins.top,
                                               margins.left,
                                               innerHeight)
                           vertices:v];

    // Right
    const CGFloat gridWidth = tState.cellConfiguration.gridSize.width * tState.cellConfiguration.cellSize.width;
    const CGFloat rightGutterWidth = tState.configuration.viewportSize.x - margins.left - margins.right - gridWidth;
    [self appendVerticesForQuad:CGRectMake(size.width - margins.right - rightGutterWidth,
                                           margins.top,
                                           margins.right + rightGutterWidth,
                                           innerHeight)
                       vertices:v];

    tState.vertexBuffer = [_verticesPool requestBufferFromContext:tState.poolContext
                                                        withBytes:vertices
                                                   checkIfChanged:YES];
}

- (int)initializeDeselected:(BOOL)deselected vertexBuffer:(iTermMarginRendererTransientState *)tState {
    CGSize size = CGSizeMake(tState.configuration.viewportSize.x,
                             tState.configuration.viewportSize.y);
    const NSEdgeInsets margins = tState.margins;
    vector_float2 vertices[6 * 8];
    vector_float2 *v = &vertices[0];
    const VT100GridRange visibleRange = VT100GridRangeMake(0, MAX(0, tState.cellConfiguration.gridSize.height));
    const VT100GridRange selectedRange = VT100GridRangeMake(tState.selectedCommandRect.origin.y,
                                                            tState.selectedCommandRect.size.height);

    int count = 0;

    // Top
    const BOOL topDeselected = tState.selectedCommandRect.origin.y > 0;
    if (deselected == topDeselected) {
        v = [self appendVerticesForQuad:CGRectMake(0,
                                                   size.height - margins.bottom,
                                                   size.width,
                                                   margins.bottom)
                               vertices:v];
        count += 1;
    }

    // Bottom
    const BOOL bottomDeselected = (!VT100GridRangeContains(selectedRange,
                                                           tState.cellConfiguration.gridSize.height) &&
                                   !tState.forceRegularBottomMargin);
    if (deselected == bottomDeselected) {
        v = [self appendVerticesForQuad:CGRectMake(0,
                                                   0,
                                                   size.width,
                                                   margins.top)
                               vertices:v];
        count += 1;
    }

    const CGFloat gridWidth = tState.cellConfiguration.gridSize.width * tState.cellConfiguration.cellSize.width;
    const CGFloat rightGutterWidth = tState.configuration.viewportSize.x - margins.left - margins.right - gridWidth;

    // Left/Right bottom deselected
    CGFloat y = margins.top;
    CGFloat h = 0;

    int lines = tState.cellConfiguration.gridSize.height - (tState.selectedCommandRect.origin.y + tState.selectedCommandRect.size.height);
    h = MAX(0, lines) * tState.cellConfiguration.cellSize.height;

    if (h > 0 && deselected) {
        v = [self appendVerticesForQuad:CGRectMake(0,
                                                   y,
                                                   margins.left,
                                                   h)
                               vertices:v];
        v = [self appendVerticesForQuad:CGRectMake(size.width - margins.right - rightGutterWidth,
                                                   y,
                                                   margins.right + rightGutterWidth,
                                                   h)
                           vertices:v];
        count += 2;
    }
    y += h;

    // Left/Right selected
    const VT100GridRange visibleSelectedRange = VT100GridRangeIntersection(visibleRange, selectedRange);
    h = visibleSelectedRange.length * tState.cellConfiguration.cellSize.height;
    if (h > 0 && !deselected) {
        v = [self appendVerticesForQuad:CGRectMake(0,
                                                   y,
                                                   margins.left,
                                                   h)
                               vertices:v];
        v = [self appendVerticesForQuad:CGRectMake(size.width - margins.right - rightGutterWidth,
                                                   y,
                                                   margins.right + rightGutterWidth,
                                                   h)
                               vertices:v];
        count += 2;
    }
    y += h;

    // Left/Right top deselected
    lines = tState.selectedCommandRect.origin.y;
    h = MAX(0, lines * tState.cellConfiguration.cellSize.height);
    if (h > 0 && deselected) {
        v = [self appendVerticesForQuad:CGRectMake(0,
                                                   y,
                                                   margins.left,
                                                   h)
                               vertices:v];
        v = [self appendVerticesForQuad:CGRectMake(size.width - margins.right - rightGutterWidth,
                                                   y,
                                                   margins.right + rightGutterWidth,
                                                   h)
                               vertices:v];
        count += 2;
    }
    y += h;

    tState.vertexBuffer = [_verticesPool requestBufferFromContext:tState.poolContext
                                                        withBytes:vertices
                                                   checkIfChanged:YES];
    return count;
}

@end

NS_ASSUME_NONNULL_END

File: sources/Metal/Renderers/iTermMarginRenderer.swift
import Metal
import simd
import AppKit

class iTermMarginRendererTransientState: iTermMetalCellRendererTransientState {
    @objc var regularVerticalColor = vector_float4(0, 0, 0, 0)
    @objc var regularHorizontalColor = vector_float4(0, 0, 0, 0)
    struct ColorExtension {
        var color: vector_float4
        var left: Bool
        var row: Int
    }
    var extensions = [ColorExtension]()
    @objc(addExtensionWithColor:left:row:)
    func addExtension(color: vector_float4, left: Bool, row: Int) {
        extensions.append(.init(color: color, left: left, row: row))
    }
}

final class iTermMarginRenderer: NSObject, iTermMetalCellRendererProtocol {
    private let blendingRenderer: iTermMetalCellRenderer
    private let nonblendingRenderer: iTermMetalCellRenderer?
    private let compositeOverRenderer: iTermMetalCellRenderer?

    private let blendingRendererForPIUs: iTermMetalCellRenderer
    private let nonblendingRendererForPIUs: iTermMetalCellRenderer?
    private let compositeOverRendererForPIUs: iTermMetalCellRenderer?

    private let colorPool: iTermMetalBufferPool
    private let fourQuadVerticesPool: iTermMetalBufferPool
    private let twoQuadVerticesPool: iTermMetalBufferPool
    private let oneQuadVerticesPool: iTermMetalBufferPool
    private let piuPool: iTermMetalMixedSizeBufferPool

    @objc(initWithDevice:)
    init(device: MTLDevice) {
        if iTermTextIsMonochrome() {
            nonblendingRenderer = iTermMetalCellRenderer(
                device: device,
                vertexFunctionName: "iTermMarginVertexShader",
                fragmentFunctionName: "iTermMarginFragmentShader",
                blending: nil,
                piuElementSize: 0,
                transientStateClass: iTermMarginRendererTransientState.self)
            compositeOverRenderer = iTermMetalCellRenderer(
                device: device,
                vertexFunctionName: "iTermMarginVertexShader",
                fragmentFunctionName: "iTermMarginFragmentShader",
                blending: iTermMetalBlending.premultipliedCompositing(),
                piuElementSize: 0,
                transientStateClass: iTermMarginRendererTransientState.self)

            nonblendingRendererForPIUs = iTermMetalCellRenderer(
                device: device,
                vertexFunctionName: "iTermMarginPIUVertexShader",
                fragmentFunctionName: "iTermMarginPIUFragmentShader",
                blending: nil,
                piuElementSize: 0,
                transientStateClass: iTermMarginRendererTransientState.self)
            compositeOverRendererForPIUs = iTermMetalCellRenderer(
                device: device,
                vertexFunctionName: "iTermMarginPIUVertexShader",
                fragmentFunctionName: "iTermMarginPIUFragmentShader",
                blending: iTermMetalBlending.premultipliedCompositing(),
                piuElementSize: 0,
                transientStateClass: iTermMarginRendererTransientState.self)
        } else {
            nonblendingRenderer = nil
            compositeOverRenderer = nil
            nonblendingRendererForPIUs = nil
            compositeOverRendererForPIUs = nil
        }

        blendingRenderer = iTermMetalCellRenderer(
            device: device,
            vertexFunctionName: "iTermMarginVertexShader",
            fragmentFunctionName: "iTermMarginFragmentShader",
            blending: iTermMetalBlending(),
            piuElementSize: 0,
            transientStateClass: iTermMarginRendererTransientState.self)!

        blendingRendererForPIUs = iTermMetalCellRenderer(
            device: device,
            vertexFunctionName: "iTermMarginPIUVertexShader",
            fragmentFunctionName: "iTermMarginPIUFragmentShader",
            blending: iTermMetalBlending(),
            piuElementSize: 0,
            transientStateClass: iTermMarginRendererTransientState.self)!

        colorPool = iTermMetalBufferPool(device: device, bufferSize: MemoryLayout<vector_float4>.stride)
        fourQuadVerticesPool = iTermMetalBufferPool(device: device, bufferSize: MemoryLayout<vector_float2>.stride * 6 * 4)
        twoQuadVerticesPool = iTermMetalBufferPool(device: device, bufferSize: MemoryLayout<vector_float2>.stride * 6 * 2)
        oneQuadVerticesPool = iTermMetalBufferPool(device: device, bufferSize: MemoryLayout<vector_float2>.stride * 6 * 1)
        piuPool = iTermMetalMixedSizeBufferPool(device: device, capacity: 512, name:"Margin PIU")

        super.init()
    }

    func createTransientStateStat() -> iTermMetalFrameDataStat {
        return .pqCreateMarginTS
    }

    func renderer(for configuration: iTermCellRenderConfiguration, forPIUs: Bool) -> iTermMetalCellRenderer {
        if forPIUs {
            if iTermTextIsMonochrome() {
                if configuration.hasBackgroundImage {
                    return compositeOverRendererForPIUs!
                } else {
                    return nonblendingRendererForPIUs!
                }
            }
            return blendingRendererForPIUs
        } else {
            if iTermTextIsMonochrome() {
                if configuration.hasBackgroundImage {
                    return compositeOverRenderer!
                } else {
                    return nonblendingRenderer!
                }
            }
            return blendingRenderer
        }
    }

    func draw(with frameData: iTermMetalFrameData,
              transientState genericState: iTermMetalCellRendererTransientState) {
        let transientState = genericState as! iTermMarginRendererTransientState
        if transientState.extensions.isEmpty {
            drawNoExtensions(with: frameData, transientState: transientState, draw: .leftAndRight)
            drawNoExtensions(with: frameData, transientState: transientState, draw: .topAndBottom)
        } else {
            drawWithExtensions(with: frameData, transientState: transientState)
        }
    }

    private func drawWithExtensions(with frameData: iTermMetalFrameData,
                                    transientState: iTermMarginRendererTransientState) {
        drawNoExtensions(with: frameData, transientState: transientState, draw: .topAndBottom)
        if transientState.configuration.viewportSize.x > transientState.configuration.viewportSizeExcludingLegacyScrollbars.x {
            drawNoExtensions(with: frameData, transientState: transientState, draw: .underLegacyScrollbar)
        }

        let cellRenderer = renderer(for: transientState.cellConfiguration, forPIUs: true)
        let size = CGSize(
            width: CGFloat(transientState.configuration.viewportSize.x),
            height: CGFloat(transientState.configuration.viewportSize.y))
        let margins = transientState.margins
        let innerHeight = size.height - margins.bottom - margins.top

        let (leftVertexBuffer, rightVertextBuffer) = initializePIUVertexBuffers(transientState: transientState)
        do {
            let leftRect = CGRect(x: 0,
                                  y: margins.top,
                                  width: margins.left,
                                  height: innerHeight)
            let (leftPIUs, count) = makePIUs(from: leftRect,
                                             bottomMargin: transientState.margins.top,
                                             extensions: transientState.extensions.filter { $0.left },
                                             lineHeight: Float(transientState.cellConfiguration.cellSize.height),
                                             defaultColor: transientState.regularVerticalColor,
                                             context: transientState.poolContext)
            transientState.pipelineState = cellRenderer.pipelineState()
            cellRenderer.draw(
                with: transientState,
                renderEncoder: frameData.renderEncoder!,
                numberOfVertices: 6,
                numberOfPIUs: count,
                vertexBuffers: [ NSNumber(value: iTermVertexInputIndexPerInstanceUniforms.rawValue): leftPIUs,
                                 NSNumber(value: iTermVertexInputIndexVertices.rawValue): leftVertexBuffer ],
                fragmentBuffers: [:],
                textures: [:])
        }

        do {
            let gridWidth = CGFloat(transientState.cellConfiguration.gridSize.width) * CGFloat(transientState.cellConfiguration.cellSize.width)
            let rightGutterWidth = CGFloat(transientState.configuration.viewportSize.x) - CGFloat(margins.left - margins.right - gridWidth)
            let rightRect = CGRect(x: size.width - margins.right - rightGutterWidth,
                                   y: margins.top,
                                   width: margins.right + rightGutterWidth,
                                   height: innerHeight)
            let (rightPIUs, count) = makePIUs(from: rightRect,
                                              bottomMargin: transientState.margins.top,
                                              extensions: transientState.extensions.filter { !$0.left },
                                              lineHeight: Float(transientState.cellConfiguration.cellSize.height),
                                              defaultColor: transientState.regularVerticalColor,
                                              context: transientState.poolContext)
            transientState.pipelineState = cellRenderer.pipelineState()
            cellRenderer.draw(
                with: transientState,
                renderEncoder: frameData.renderEncoder!,
                numberOfVertices: 6,
                numberOfPIUs: count,
                vertexBuffers: [ NSNumber(value: iTermVertexInputIndexPerInstanceUniforms.rawValue): rightPIUs,
                                 NSNumber(value: iTermVertexInputIndexVertices.rawValue): rightVertextBuffer ],
                fragmentBuffers: [:],
                textures: [:])
        }
    }

    private func makePIUs(from container: NSRect,
                          bottomMargin: CGFloat,
                          extensions: [iTermMarginRendererTransientState.ColorExtension],
                          lineHeight: Float,
                          defaultColor: vector_float4,
                          context: iTermMetalBufferPoolContext) -> (MTLBuffer, Int) {
        let sorted = extensions.sorted { $0.row < $1.row }
        var result = Array<iTermMarginExtensionPIU>()
        for e in sorted {
            var color = e.color
            if iTermTextIsMonochrome() {
                color.x *= color.w
                color.y *= color.w
                color.z *= color.w
            }
            result.append(.init(
                color: color,
                defaultBackgroundColor: defaultColor,
                yOffset: Float(container.height) - Float(e.row + 1) * lineHeight + Float(bottomMargin)))
        }
        let buffer = result.withUnsafeBytes { urbp in
            let pius = urbp.baseAddress!
            return piuPool.requestBuffer(from: context,
                                         size: result.count * MemoryLayout<iTermMarginExtensionPIU>.stride,
                                         bytes: pius)
        }
        return (buffer, result.count)
    }

    private func drawNoExtensions(with frameData: iTermMetalFrameData,
                                  transientState: iTermMarginRendererTransientState,
                                  draw: Draw) {
        initializeRegularVertexBuffer(tState: transientState, draw: draw)
        var color = transientState.regularHorizontalColor
        if iTermTextIsMonochrome() {
            color.x *= color.w
            color.y *= color.w
            color.z *= color.w
        }
        let colorBuffer = colorPool.requestBuffer(
            from: transientState.poolContext,
            withBytes: &color,
            checkIfChanged: true)
        let cellRenderer = renderer(for: transientState.cellConfiguration, forPIUs: false)
        transientState.pipelineState = cellRenderer.pipelineState()
        cellRenderer.draw(
            with: transientState,
            renderEncoder: frameData.renderEncoder!,
            numberOfVertices: 6 * 4,
            numberOfPIUs: 0,
            vertexBuffers: [ NSNumber(value: iTermVertexInputIndexVertices.rawValue): transientState.vertexBuffer ],
            fragmentBuffers: [ NSNumber(value: iTermFragmentBufferIndexMarginColor.rawValue): colorBuffer ],
            textures: [:])
    }

    var rendererDisabled: Bool {
        return false
    }

    func createTransientState(forCellConfiguration configuration: iTermCellRenderConfiguration,
                              commandBuffer: MTLCommandBuffer) -> iTermMetalRendererTransientState? {
        let renderer = renderer(for: configuration, forPIUs: false)
        let ts = renderer.createTransientState(
            forCellConfiguration: configuration,
            commandBuffer: commandBuffer)!
        return ts
    }

    @discardableResult
    private func appendVerticesForQuad(
        _ quad: CGRect,
        to v: UnsafeMutablePointer<vector_float2>) -> UnsafeMutablePointer<vector_float2> {
        v.pointee = simd_make_float2(Float(quad.maxX), Float(quad.minY))
        let v1 = v.advanced(by: 1)
        v1.pointee = simd_make_float2(Float(quad.minX), Float(quad.minY))
        let v2 = v.advanced(by: 2)
        v2.pointee = simd_make_float2(Float(quad.minX), Float(quad.maxY))
        let v3 = v.advanced(by: 3)
        v3.pointee = simd_make_float2(Float(quad.maxX), Float(quad.minY))
        let v4 = v.advanced(by: 4)
        v4.pointee = simd_make_float2(Float(quad.minX), Float(quad.maxY))
        let v5 = v.advanced(by: 5)
        v5.pointee = simd_make_float2(Float(quad.maxX), Float(quad.maxY))
        return v.advanced(by: 6)
    }

    private func initializePIUVertexBuffers(transientState tState: iTermMarginRendererTransientState) -> (MTLBuffer, MTLBuffer) {
        let margins = tState.margins
        var leftVertices = [vector_float2](repeating: .zero, count: 6)
        leftVertices.withUnsafeMutableBufferPointer { buf in
            var v = buf.baseAddress!
            v = appendVerticesForQuad(
                CGRect(x: 0,
                       y: 0,
                       width: margins.left,
                       height: tState.cellConfiguration.cellSize.height),
                to: v)
        }
        var rightVertices = [vector_float2](repeating: .zero, count: 6)
        rightVertices.withUnsafeMutableBufferPointer { buf in
            var v = buf.baseAddress!
            let gridWidth = CGFloat(tState.cellConfiguration.gridSize.width) * CGFloat(tState.cellConfiguration.cellSize.width)
            let rightGutterWidth = CGFloat(tState.configuration.viewportSizeExcludingLegacyScrollbars.x) - margins.left - margins.right - gridWidth
            let width = margins.right + rightGutterWidth
            v = appendVerticesForQuad(
                CGRect(x: CGFloat(tState.configuration.viewportSizeExcludingLegacyScrollbars.x) - width,
                       y: 0,
                       width: width,
                       height: tState.cellConfiguration.cellSize.height),
                to: v)
        }
        return (
            oneQuadVerticesPool.requestBuffer(
                from: tState.poolContext,
                withBytes: leftVertices,
                checkIfChanged: true),
            oneQuadVerticesPool.requestBuffer(
                from: tState.poolContext,
                withBytes: rightVertices,
                checkIfChanged: true)
        )
    }

    private enum Draw {
        case topAndBottom
        case leftAndRight
        case underLegacyScrollbar
    }
    private func initializeRegularVertexBuffer(tState: iTermMarginRendererTransientState,
                                               draw: Draw) {
        let size = CGSize(
            width: CGFloat(tState.configuration.viewportSize.x),
            height: CGFloat(tState.configuration.viewportSize.y))
        let margins = tState.margins
        var vertices = [vector_float2](repeating: .zero, count: 6 * 2)
        vertices.withUnsafeMutableBufferPointer { buf in
            var v = buf.baseAddress!
            switch draw {
            case .topAndBottom:
                // Top
                v = appendVerticesForQuad(
                    CGRect(x: 0,
                           y: 0,
                           width: size.width,
                           height: margins.top),
                    to: v)
                // Bottom
                v = appendVerticesForQuad(
                    CGRect(x: 0,
                           y: size.height - margins.bottom,
                           width: size.width,
                           height: margins.bottom),
                    to: v)
            case .leftAndRight:
                let innerHeight = size.height - margins.bottom - margins.top
                // Left
                v = appendVerticesForQuad(
                    CGRect(x: 0,
                           y: margins.top,
                           width: margins.left,
                           height: innerHeight),
                    to: v)
                // Right
                let gridWidth = CGFloat(tState.cellConfiguration.gridSize.width) * CGFloat(tState.cellConfiguration.cellSize.width)
                let rightGutterWidth = CGFloat(tState.configuration.viewportSize.x) - margins.left - margins.right - gridWidth
                v = appendVerticesForQuad(
                    CGRect(x: size.width - margins.right - rightGutterWidth,
                           y: margins.top,
                           width: margins.right + rightGutterWidth,
                           height: innerHeight),
                    to: v)
            case .underLegacyScrollbar:
                let innerHeight = size.height - margins.bottom - margins.top
                v = appendVerticesForQuad(
                    CGRect(x: CGFloat(tState.configuration.viewportSizeExcludingLegacyScrollbars.x),
                           y: margins.top,
                           width: CGFloat(tState.configuration.viewportSize.x - tState.configuration.viewportSizeExcludingLegacyScrollbars.x),
                           height: innerHeight),
                    to: v)
            }
        }
        tState.vertexBuffer = twoQuadVerticesPool.requestBuffer(
            from: tState.poolContext,
            withBytes: vertices,
            checkIfChanged: true)
    }

    @discardableResult
    func initializeVertexBuffer(
        _ tState: iTermMarginRendererTransientState) -> Int {
        let size = CGSize(
            width: CGFloat(tState.configuration.viewportSize.x),
            height: CGFloat(tState.configuration.viewportSize.y))
        let margins = tState.margins
        var vertices = [vector_float2](repeating: .zero, count: 6 * 8)
        var count = 0
        vertices.withUnsafeMutableBufferPointer { buf in
            var v = buf.baseAddress!
            // Top
            v = appendVerticesForQuad(
                CGRect(x: 0,
                       y: size.height - margins.bottom,
                       width: size.width,
                       height: margins.bottom),
                to: v)
            count += 1
            // Bottom
            v = appendVerticesForQuad(
                CGRect(x: 0,
                       y: 0,
                       width: size.width,
                       height: margins.top),
                to: v)
            count += 1
            let gridWidth = CGFloat(tState.cellConfiguration.gridSize.width) * CGFloat(tState.cellConfiguration.cellSize.width)
            let rightGutterWidth = CGFloat(tState.configuration.viewportSize.x) - margins.left - margins.right - gridWidth
            let y = margins.top
            let h = CGFloat(VT100GridRangeMake(0, max(0, tState.cellConfiguration.gridSize.height)).length)
                * tState.cellConfiguration.cellSize.height
            v = appendVerticesForQuad(
                CGRect(x: 0,
                       y: y,
                       width: margins.left,
                       height: h),
                to: v)
            v = appendVerticesForQuad(
                CGRect(x: size.width - margins.right - rightGutterWidth,
                       y: y,
                       width: margins.right + rightGutterWidth,
                       height: h),
                to: v)
            count += 2
        }
        tState.vertexBuffer = fourQuadVerticesPool.requestBuffer(
            from: tState.poolContext,
            withBytes: vertices,
            checkIfChanged: true)
        return count
    }
}
